src/ray/raylet/runtime_env_agent_client.cc

```diff
@@ -0,0 +1,395 @@
+            callback(true,
+                     reply.serialized_runtime_env_context(),
+                     /*setup_error_message*/ "");
+          }
+        },
+        /*fail_callback=*/
+        [=](ray::Status status) {
+          RAY_LOG(INFO)
+              << "Failed to create runtime env for job " << job_id
+              << ", status = " << status
+              << ", maybe there are some network problems, will fail the request.";
+          RAY_LOG(DEBUG) << "Serialized runtime env for job " << job_id << ": "
+                         << serialized_runtime_env;
+          callback(false, "", "Failed to request agent.");
+        });
+  }
+
+  // Does the real work of calling HTTP.
+  // Returns {error, empty} or {ok, reply}.
+  void TryGetOrCreateRuntimeEnv(
+      const JobID &job_id,
+      const std::string &serialized_runtime_env,
+      const rpc::RuntimeEnvConfig &runtime_env_config,
+      const std::string &serialized_allocated_resource_instances,
+      std::function<void(rpc::GetOrCreateRuntimeEnvReply)> succ_callback,
+      std::function<void(ray::Status)> fail_callback) {
+    rpc::GetOrCreateRuntimeEnvRequest request;
+    request.set_job_id(job_id.Hex());
+    request.set_serialized_runtime_env(serialized_runtime_env);
+    request.mutable_runtime_env_config()->CopyFrom(runtime_env_config);
+    request.set_serialized_allocated_resource_instances(
+        serialized_allocated_resource_instances);
+    std::string payload = request.SerializeAsString();
+
+    auto session = Session::Create(
+        io_context_,
+        address_,
+        port_str_,
+        "/get_or_create_runtime_env",
+        std::move(payload),
+        /*succ_callback=*/
+        [succ_callback, fail_callback](std::string body) {
+          rpc::GetOrCreateRuntimeEnvReply reply;
+          if (!reply.ParseFromString(body)) {
+            fail_callback(Status::IOError("protobuf parse error"));
+          } else {
+            succ_callback(std::move(reply));
+          }
+        },
+        fail_callback);
+    session->run();
+  }
+
+  // Making HTTP call.
+  // POST /delete_runtime_env_if_possible
+  // Body = proto rpc::DeleteRuntimeEnvIfPossibleRequest
+  void DeleteRuntimeEnvIfPossible(const std::string &serialized_runtime_env,
+                                  DeleteRuntimeEnvIfPossibleCallback callback) override {
+    RetryInvoke<rpc::DeleteRuntimeEnvIfPossibleReply>(
+        [=](SuccCallback<rpc::DeleteRuntimeEnvIfPossibleReply> succ_callback,
+            FailCallback fail_callback) {
+          return TryDeleteRuntimeEnvIfPossible(
+              serialized_runtime_env, succ_callback, fail_callback);
+        },
+        /*succ_callback=*/
+        [=](rpc::DeleteRuntimeEnvIfPossibleReply reply) {
+          if (reply.status() != rpc::AGENT_RPC_STATUS_OK) {
+            // TODO(sang): Find a better way to delivering error messages in this
+            RAY_LOG(ERROR) << "Failed to delete runtime env "
```

Comment:
Can you make it WARNING?

It is a bit weird, but all the ERROR logs are printed to user driver, so we'd like to avoid doing that unless we want to expose this error message to users' driver scripts